lib/Parse/ParseExpr.cpp

```diff
@@ -3007,6 +3016,90 @@ bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
   return IsError;
 }
 
+ExprResult Parser::ParseBoundsCastExpression(IdentifierInfo &Ident,
+                                             SourceLocation &ILoc) {
+  ExprResult Result(true);
+  ParsedType CastTy = nullptr;
+  BoundsCastExpr::Kind kind;
+  if (&Ident == Ident_dynamic_bounds_cast)
+    kind = BoundsCastExpr::Kind::Dynamic;
+  else if (&Ident == Ident_assume_bounds_cast)
+    kind = BoundsCastExpr::Kind::Assume;
+  else
+    kind = BoundsCastExpr::Kind::Invalid;
+
+  DeclSpec DS(AttrFactory);
+
+  if (ExpectAndConsume(tok::less, diag::err_expected_less_after,
+                       Ident.getNameStart()))
+    return ExprError();
+
+  ParseSpecifierQualifierList(DS);
```

Comment:
A type name in C is not what you might expect it to be.   It is actually the name of a type, not just an identifier.  See Section 6.7.7 of the C11 specification.    You should be calling ParseTypeName and then checking for a comma.  You should be calling ParseSpecifierQualifierList, checking for a comma, and then parsing a declarator.